tests/protocol/test_protocol_util.py

```diff
@@ -17,18 +17,55 @@
 
 import os
 import unittest
+from multiprocessing import Queue
+from threading import Thread
 
 from tests.tools import AgentTestCase, MagicMock, Mock, patch
 from azurelinuxagent.common.exception import *
 from azurelinuxagent.common.protocol import get_protocol_util, \
                                             TAG_FILE_NAME
 from azurelinuxagent.common.utils.restutil import KNOWN_WIRESERVER_IP
-from azurelinuxagent.common.protocol.util import ENDPOINT_FILE_NAME
+from azurelinuxagent.common.protocol.util import ProtocolUtil
 from errno import ENOENT
 
 
 @patch("time.sleep")
 class TestProtocolUtil(AgentTestCase):
+
+    def setUp(self):
+        super(TestProtocolUtil, self).setUp()
+        ProtocolUtil.clear()
+
+    def test_get_protocol_util_should_return_same_object_for_same_thread(self, _):
+        protocol_util1 = get_protocol_util()
+        protocol_util2 = get_protocol_util()
+
+        self.assertEqual(protocol_util1, protocol_util2)
+
+    def test_get_protocol_util_should_return_different_object_for_different_thread(self, _):
+        def get_util_obj(q, err):
+            try:
+                q.put(get_protocol_util())
+            except Exception as e:
+                err.put(str(e))
+
+        queue = Queue()
+        errors = Queue()
+        t1 = Thread(target=get_util_obj, args=(queue, errors))
+        t2 = Thread(target=get_util_obj, args=(queue, errors))
+        t1.start()
```

Comment:
nit: for extra checks, maybe consider giving the threads different names and asserting that the names match the respective singletons.